Add an example of usage for the OSRD data formats (#28)
DGEXSolutions/osrd-website: content/developers/internals/data-model/usage-example.en.md

---
title: "Usage Example"
linkTitle: "Example"
weight: 20

> description: "OSRD data format usage example"

Author: I suggested that but @flomonster said it was fine in one article

> description: "OSRD data format usage example"

Reviewer: alright, lets reconsider later

---

## Introduction

This page gives an example of how the data formats are used to describe an infrastructure in *OSRD*.

For this purpose, let's take as an example the following toy infrastructure:

![Toy infrastructure schema](../svg_schemas/small_infra_schema.drawio.en.svg)

{{% alert title="Tip" color="info"%}}
In case you want to zoom in on schemas, you can click on the little editing button that appears when hovering the mouse on it.
{{% /alert %}}

> In this schema we represent the overall look of the infrastructure with just the lines and the stations.

Author: ```suggestion
In this diagram we give an overview of the infrastructure with just lines and stations.
```

This is obviously not a realistic infrastructure, but it will allow us to present the data models in a simple way.

> 

Author: There are some things that are useless, like the loop to go from north to south or the double cross switch, which is rarely used like that.
Also the lengths of the track sections are meh.
Does it not make it an unrealistic infrastructure to you ?

> 

Reviewer: alright, how about "This infrastructure is not meant to be realistic, but rather the help illustrate OSRD's data model"

We will build the details of the infrastructure step by step, and present the objects used at each step.

### The infrastructure generator

> In the [*OSRD* repository](https://github.com/DGEXSolutions/osrd) is a [python module](https://github.com/DGEXSolutions/osrd/tree/dev/core/examples/generated) that allows for generating infrastructures in a format understood by *OSRD*.

Author: ```suggestion
In the [*OSRD* repository](https://github.com/DGEXSolutions/osrd) is a [python library](https://github.com/DGEXSolutions/osrd/tree/dev/core/examples/generated) designed to help generate infrastructures in a format understood by *OSRD*.
```

> The infrastructure we talk about in this page can be generated through the [small_infra.py](https://github.com/DGEXSolutions/osrd/blob/dev/core/examples/generated/scripts/small_infra.py) file. For more information about the generation scripts, you can check out the related [README](https://github.com/DGEXSolutions/osrd/blob/dev/core/examples/generated/README.md).

Author: ```suggestion
The infrastructure discussed in this section can be generated thanks to [small_infra.py](https://github.com/DGEXSolutions/osrd/blob/dev/core/examples/generated/scripts/small_infra.py) file. To learn more about the generation scripts, you can check out the related [README](https://github.com/DGEXSolutions/osrd/blob/dev/core/examples/generated/README.md).
```

> ## The rails

Author: I fear that this might be confusing for newcomers, or maybe if we call the section graph, we could add a paragraph about why an infrastructure can be seen as a graph ?

### Track sections

The first objects that we need to define are the `TrackSections`. Most other objects are defined in relation to the track sections we have set up.

> A track section is simply a portion of rail (switches not included). One can chose to divide the tracks of their infrastructure in as many track sections as they like. Here we chose to use the longest track sections possible, which means that between two switches there is always a single track section.

Author: ```suggestion
A track section is simply a section of rail (switches not included). One can chose to divide the tracks of their infrastructure in as many track sections as they like. Here we chose to use the longest track sections possible, which means that between two switches there is always a single track section.
```

> A train runs on one track section, if you want to run trains in opposite directions at the same time and on the same line, you need to define at least two parallel track sections.

Author: Fine.
I wanted to point out that a track section really is just one track and not the two tracks that we see most often when taking trains in IDF for example. 
I had a little bit of confusion with that when making `small_infra.py` so I thought other could have that kind of confusion too.
```suggestion
Track sections is what simulated trains roll onto. They are the abstract equivalent to physical rail sections. Track sections are bidirectional.
```

For our example, we will define two tracks for the line between the West and North-East stations.

> Furthermore, we will add overpassing tracks at the North and Mid-West stations, to make our infrastructure a bit more realistic.

Author: ```suggestion
We also have overpassing tracks at the North and Mid-West stations for added realism.
```

> We will also have three different tracks in the West station, since it's a major hub in our imaginary infrastructure.

Author: ```suggestion
We will also have three separate tracks in the West station, since it's a major hub in our imaginary infrastructure.
```

* `sch`: the coordinates in the schematic view (sch for schematic), also in [GeoJSON](https://en.wikipedia.org/wiki/GeoJSON) format.
* cosmetic attributes: `line_name`, `track_name`, `track_number` which are used to indicate the name and labels that were given to the tracks / lines in real life.

For all track sections in our infrastructure, the `geo` and `sch` attributes are identical, and very much resemble the given schema.

For most track sections, their `length` is proportional to what can be seen in the schema, but for readability, we made *TA6*, *TA7*, *TD0* and *TD1* not correspond to their length (which are 10km and 25km).

### Track Section Links

For the moment we only created track sections, they have length and coordinates but they are not linked to each other (we do not check which coordinates overlap, thank you very much). We simply indicate the link between two track sections by creating an object, linking one extremity to the other.

If two track sections are not connected by a `TrackSectionLink`, a train in an *OSRD* simulation cannot go from the one to the other.

> In our example, creating the links is straight-forward: since we chose to have long track sections, we only have to link them at switches. Happily, in our generation scripts, like in [small_infra.py](https://github.com/DGEXSolutions/osrd/blob/dev/core/examples/generated/scripts/small_infra.py), when creating switches, the corresponding `TrackSectionLink`s are automatically created.

Reviewer 2: This isn't true. We don't generate track sections link when we create a switch.

Track sections link are only useful to link two tracks.

Reviewer 2: My bad I didn't checked the code. This is consider as a warning to have links overlapping a switch.

So we are generated infrastructures with warnings...

Author: Oh so switches acts as three (or more)-way links ?

Let's delve a bit on the switches.

### Switches

<!-- TODO: what it is -->
The switches are the parts needed to link more than two different track sections together.

> A switch manages the different links that are associated to it (always at least two). In general, it translates into the following rule: when one link is used, the others cannot be used at the same time. This is because if two links were used by trains at the same time, it would be an accident.

Reviewer 2: > (always at least two)

We can define a SwitchType with only one port (and it makes sense for cross switches.

Author: Same question as in [comment below](https://github.com/DGEXSolutions/osrd-website/pull/28#discussion_r899891960), I think I don't really understand the use of groups

> Since the kinds of links that we need to create at a switch often happen to be redundant, we have `SwitchType` objects, which gather what links should be created, when creating this particular type of switch. We will describe this object in the next subsection.

Reviewer 2: I'm not a big fan of talking about link here.

We should rather talk about the different configurations (groups) of a switch.

<!-- TODO: how it's implemented -->
<!-- TODO: how we use it -->

#### Switch Types

A switch type contains two attributes:

* `ports`: A list of names, that will be used to reference the ports. It is best described by example (see next paragraph).

> * `groups`: A mapping from group names to lists of connections between two ports. When one group is active, only the connections in its list are active.

Reviewer 2: Same here. You can talk of configuration or state of the switch.

In *OSRD*, we currently use three switch types.

**1) Point Switch**

This is the most standard kind of switch and can be conceived as two tracks merging into one, or one track splitting into two.

This switch type has three ports: *base*, *left* and *right*.

![Point switch schema](../svg_schemas/point_switch.en.svg)

There are two groups, each with one connection in their list: `LEFT`, which links *base* to *left*, and `RIGHT` which links *base* to *right*.

Thus, at any given moment, a train can be able to go from *base* to *left* or from *base* to *right* but never to both at the same time. A train cannot go from *left* to *right*.

A Point Switch only has two positions:

![Point switch positions schema](../svg_schemas/point_switch_positions.en.svg)

**2) Cross Switch**

This is simply two tracks crossing each other.

This type has four ports: *north*, *south*, *east* and *west*.

![Cross switch schema](../svg_schemas/cross_switch.en.svg)

It has two groups containing each one connection: *north* to *south* and *west* to *east*. Thus the cross switch has two positions, as shown here:

> ![Cross switch positions schema](../svg_schemas/cross_switch_positions.en.svg)

Reviewer 2: A cross switch has only one group:
```json
    {
        "id": "cross",
        "ports": [
            "north",
            "south",
            "east",
            "west"
        ],
        "groups": {
            "static": [
                {
                    "dst": "west",
                    "src": "east",
                    "bidirectional": true
                },
                {
                    "dst": "south",
                    "src": "north",
                    "bidirectional": true
                }
            ]
        }
    }
```

Author: @multun told me that this type of switch should have two groups and not one, so I chose to document the "normal" situation, but I can document the one group documentation if your prefer. However, first I need to ask you questions about how groups work because then I did not understand it correctly.

Reviewer 2: We can check with @loic-hamelin or @nicolaswurtz but I am pretty sure that this type of switch is passive.

## The rails

### Track sections

The first objects that we need to define are the `TrackSections`. Most other objects are defined in relation to the track sections we have set up.

A track section is simply a portion of rail (switches not included). One can chose to divide the tracks of their infrastructure in as many track sections as they like. Here we chose to use the longest track sections possible, which means that between two switches there is always a single track section.

A train runs on one track section, if you want to run trains in opposite directions at the same time and on the same line, you need to define at least two parallel track sections.

For our example, we will define two tracks for the line between the West and North-East stations.

Furthermore, we will add overpassing tracks at the North and Mid-West stations, to make our infrastructure a bit more realistic.

We will also have three different tracks in the West station, since it's a major hub in our imaginary infrastructure.

Here are the track sections that we then need:

![Track sections schema](../svg_schemas/small_infra_rails.drawio.en.svg)

{{% alert title="Important" color="warning"%}}

> They are represented as arrows in our schema to stress the fact that they have a **start** and an **end**. This is important because the objects related to the track section are positioned using an offset, and the offset is a distance from the **start** of the track section.

Author: ```suggestion
`TrackSections` are represented as arrows in this diagram to stress the fact that they have a **start** and an **end**. This is important because the objects related to the track section are positioned using an offset, and the offset is a distance from the **start** of the track section.
```

> Therefore, if you want to place an object at the beginning of the track section, you would set its offset to 0. To place it at the end of the track section, its offset should be the `length` of the track section.

Reviewer: I think it's just fine but as you wish, it works too. beware, it should be "its offset would need to be" and not "it's offset would need to be" (notice the missing apostrophe)

{{% /alert %}}

For the objects to be complete, there are different attributes to set:

* `length`: the length of the track section in meters.
* `geo`: the coordinates in real life (geo is for geographic), in the [GeoJSON](https://en.wikipedia.org/wiki/GeoJSON) format.
* `sch`: the coordinates in the schematic view (sch for schematic), also in [GeoJSON](https://en.wikipedia.org/wiki/GeoJSON) format.
* cosmetic attributes: `line_name`, `track_name`, `track_number` which are used to indicate the name and labels that were given to the tracks / lines in real life.

For all track sections in our infrastructure, the `geo` and `sch` attributes are identical, and very much resemble the given schema.

For most track sections, their `length` is proportional to what can be seen in the schema, but for readability, we made *TA6*, *TA7*, *TD0* and *TD1* not correspond to their length (which are 10km and 25km).

### Track Section Links

For the moment we only created track sections, they have length and coordinates but they are not connected to each other (we do not check which coordinates overlap, thank you very much).

> If you only need to connect two `TackSections` together, this is where `TrackSectionLinks` come into play. A train can go from one track section to the other only if they are connected by a `TrackSectionLink` (or by a `Switch`).

Author: ```suggestion
If you only need to connect two `TrackSections` together, this is where `TrackSectionLinks` come into play. `TrackSectionLinks` are used to connect two track sections together, just like a weld joint would in real life. In an OSRD simulation, a train can go from one track section to the other only if they are connected by a `TrackSectionLink` (or by a `Switch`).
```

If you want need to connect more `TackSections` together, you will need the [`Switches`](#switches) objects.

In our infrastructure, since we chose to have our track sections as long as possible, we do not actually need to use `TrackSectionLinks`.

### Switches

A `Switch` object can be understood as a manageable superposition of `TrackSectionLinks`: when instantiating the `Switch`, you define all the connections that you want, and in the *OSRD* core, all these connections will be taken into account, however only one can be active at the same time. Indeed, since we are talking about some sort of overlap of rail, if two connections are activated at once, then two trains could want to use them at the same time, and that would result in a collision (not head on, but side to side, in french called *en écharpe*).

> In general, in an infrastructure, switches more or less use the same connection layouts, so in accordance with [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself), we have `SwitchType` objects, which gather what connections should be created, when creating this particular type of switch. We describe this object in the next subsection.

Author: ```suggestion
In an infrastructure, switches more or less use the same connection layouts so, to avoid repetitions, we have `SwitchType` objects, which gather what connections should be created, when creating this particular type of switch. We describe this object in the next subsection.
```

#### Switch Types

> A `SwitchType` contains two attributes:

Author: ```suggestion
`SwitchTypes` have two mandatory attributes:
```

* `ports`: A list of names, that will be used to reference the ports. It is best described by example (see next paragraph).
* `groups`: A mapping from group names to lists of connections between two ports.

> At any time, only one group can ever be active, and among the connections of this group, only one can be active. During a simulation, it is instantaneous to change the activated connection inside a group, however, it can take some time to change the activate group.  

Author: ```suggestion
At any time, only one group can ever be active, and among the connections of this group, only one can be active in a given direction. During a simulation, it is instantaneous to change the activated connection inside a group. Changing the activated group takes a configurable time.
```

This is because a switch is a physical object, and changing connections can imply moving parts of the objects. `Groups` are designed to represent the different positions that a switch can have. Inside one `group` are all the connections that can be accessed in the associated position of the switch.

The duration needed to change from one group to the other is stored inside the `Switch`, since this duration can vary depending on the physical implementation of the switch.

> In *OSRD*, we currently use three `SwitchTypes`.

Author: ```suggestion
In *OSRD*, we currently use three `SwitchTypes` instances, but more can be added in order to suit an infrastructure better.
```

**1) The Point Switch**

> This is the most standard kind of switch and can be conceived as two tracks merging into one, or one track splitting into two.

Author: ```suggestion
The ubiquitous Y switch, which can be thought of as either two tracks merging, or one track splitting.
```

### Track Section Links

For the moment we only created track sections, they have length and coordinates but they are not connected to each other (geospacial data is not used to deduce which tracks connect).

If you only need to connect two `TrackSections` together, this is where `TrackSectionLinks` come into play. `TrackSectionLinks` are used to connect two track sections together, just like a weld joint would in real life. In an OSRD simulation, a train can go from one track section to the other only if they are connected by a `TrackSectionLink` (or by a `Switch`).

If you need to connect more `TrackSections` together, use the [`Switches`](#switches) objects.

In our infrastructure, since we chose to have our track sections as long as possible, we do not actually need to use `TrackSectionLinks`.
`TrackSectionLinks` are always optional: two track sections connected by a link behaves just like a single track section.

### Switches

A `Switch` represents just what you would expect: railway switches.

> The way a `Switch` object is designed in OSRD can be understood as a manageable superposition of `TrackSectionLinks`. When instantiating the `Switch`, you define all the connections that you want, and in the *OSRD* core, all these connections will be taken into account. However, only one can be active at the same time. Indeed, we are talking about some sort of overlap of rail. If two connections are activated at once, then two trains could want to use them at the same time, and that would result in a collision (not head on, but side to side, in french called *en écharpe*).

Reviewer: it's just a way of understanding it

> In an infrastructure, switches more or less use the same connection layouts so, to avoid repetitions, we have `SwitchType` objects, which gather what connections should be created, when creating this particular type of switch. We describe this object in the next subsection.

Reviewer: I've updated my suggestion

The slope / curve value that one wants to set is constant on the whole range. For varying curves / slopes, one needs to create several objects.

{{% alert title="" color="primary"%}}
Mind that the `begin` value should always be smaller than the `end` value. That is why the curve / slope values can be negative: an uphill slope of 1 going from offset 10 to 0 is the same as a downhill slope of -1 going from offsets 0 to 10.
{{% /alert %}}

In the [small_infra.py](https://github.com/DGEXSolutions/osrd/blob/dev/core/examples/generated/scripts/small_infra.py) file, we have slopes on the track sections *TA6*, *TA7*, *TD0* and *TD1*.

There are curves as well, on the track sections *TE0*, *TE1*, *TE3* and *TF1*.

### Operational points

`OperationalPoints` represent anything that we find of interest and want to represent in our simulation. An operational point is mostly constituted of the list of its parts, which are simply positions on track sections.

For example, we might find convenient to have the position of the platforms actually integrated in our system. For that purpose, we can represent them as operational points.

> In our example infrastructure, they are the only operational points used. The operational point parts are represented by purple diamonds in the following schema, the operational points they are related to are simply the names they are close by.

Author: ```suggestion
In the example infrastructure, we only used operational points to represent stations. Operational point parts are displayed as purple diamonds. Keep in mind a single operational point may contain multiple parts.
```

![Operational points examples](../svg_schemas/operational_points.drawio.en.svg)

## Other objects

> Now we have represented all the material parts of our infrastructure <span style="color: red">really ?</span>. But there are still concepts that we may want to represent.

Reviewer: ```suggestion
```

### Loading Gauge Limits

These objects are akin to `Slopes` and `Curves`: it covers a range of track section, with a `begin` and an `end` offset. It represents a restriction on the trains that can travel on the given range, by weight or by train type (freight or passenger).

We did not put any in our examples.

### Speed Sections

The `SpeedSections` represent speed limitations (in meters per second) that are applied on some parts of the tracks. One `SpeedSection` can span on several track sections, and do not necessarily cover the whole track sections. Speed sections can overlap.

In our example infrastructure, we have a speed section covering the whole infrastructure, limiting the speed to 300 km/h. On a smaller part of the infrastructure, we applied more restrictive speed sections. They are represented on the next image.

![Speed section examples](../svg_schemas/speed_sections.en.svg)

### Routes

> A `Route` in OSRD represents a trajectory a train take. It is represented with the following attributes:

Reviewer: this is part of interlocking, and pretty important: a route is not a whole train path, only a fragment. It's an allowed move: without a route, trains can't go anywhere.
it should be moved to `Interlocking`

A track section is simply a section of rail (switches not included). One can chose to divide the tracks of their infrastructure in as many track sections as they like. Here we chose to use the longest track sections possible, which means that between two switches there is always a single track section.

Track sections is what simulated trains roll onto. They are the abstract equivalent to physical rail sections. Track sections are bidirectional.

{{% alert title="" color="info"%}}
**N.B. :** A track section does not represent part of a line, but really part of a track.
{{% /alert %}}

In this example, we define two tracks for the line between the West and North-East stations. We also have overpassing tracks at the North and Mid-West stations for added realism. Finally, we have three separate tracks in the West station, since it's a major hub in our imaginary infrastructure.

Here are the track sections that we then need:

![Track sections diagram](../svg_diagrams/small_infra_rails.drawio.en.svg)

{{% alert title="Important" color="warning"%}}

> `TrackSections` are represented as arrows in this diagram to stress the fact that they have a **start** and an **end**. This is important because the objects related to the track section are positioned using an offset, and the offset is a distance from the **start** of the track section.

Author: ```suggestion
`TrackSections` are represented as arrows in this diagram to stress the fact that they have a **start** and an **end**. It matters as objects positioned on track sections are located using their distance from the **start** of their track section.
```

Here you can see all the spots where we placed detectors in our infrastructure, represented by green squares. We need not include their `id` for readability.

![Infra diagram with all detectors](../svg_diagrams/small_infra_detectors.drawio.en.svg)

{{% alert title="" color="info"%}}
The clumped up squares represent many detectors at once. Indeed, because some track sections are not represented with their full length, we could not represent all the detectors on the corresponding track section.
{{% /alert %}}

**N.B. :**

* Between some points, we added only one detector (and not two), because they were really close together, and it would have made no sense to create a tiny TVDS between the two. This situation happened on track sections (*TA3*, *TA4*, *TA5*, *TF0* and *TG3*).
* In our infrastructure, there is relatively few track sections which are long enough to require more detectors than just those related to switches. Namely, *TA6*, *TA7*, *TDO*, *TD1*, *TF1*, *TG1* and *TH1*. For example  *TD0*, which measures 25km, has in fact 17 detectors in total.

### Buffer stops

> The `BufferStops` are physical objects that prevent trains from leaving the rails.

Reviewer: ```suggestion
`BufferStops` are obstacles designed to prevent trains from sliding off dead ends.
```

### Routes

A `Route` in OSRD represents an itinerary in the infrastructure. A train's path can be composed of several `Route` objects. They are used when interacting with path reservation for trains and are thus part of the interlocking system. For more information on routing and reservations see the dedicated pages ([link](/fr/developers/internals/simulation-model/)).

It is represented with the following attributes:

* `entry_point` and `exit_point`: in *OSRD* those are generally references to detectors or buffer stops.
* `path`: The list of track section ranges a train can pass on to go from entry to exit. There can be two Routes with the same entry and exit points and two different paths.
* `release_detectors`: The list of detectors that can be released after the train has crossed them. This allows for "hard" reservation when the list is empty and "soft" when the list contains all the detectors seen along the way.

In our infrastructure, all possible routes were generated automatically by the script.

## Signaling

> Thanks to interlocking, trains are located and allowed to move. It's a good start, but meaningless until trains are made aware of it. This is where the `Signal`s come into play: signals react to interlocking, and can be seen by trains.

Reviewer: ```suggestion
Thanks to interlocking, trains are located and allowed to move. It's a good start, but meaningless until trains are made aware of it. This is where `Signal`s come into play: signals react to interlocking, and can be seen by trains.
```

> How trains react to signals depends on signal type.

Reviewer: ```suggestion
How trains react to signals depends on the aspect, kind of signal, and signaling system.
```

> Here are the most important attributes for the signals:

Reviewer: ```suggestion
Here are the most important attributes for signals:
```

> * `linked_detector`: The detector it is linked to.

Reviewer: ```suggestion
* `linked_detector`: The linked detector.
```

* `type_code`: The type of signal.
* `direction`: The direction it protects, which can simply be interpreted as the way in which it can be seen by an incoming train (since there are lights only on one side...).
* Cosmetic attributes like `angle_geo` or `side` which control the way in which the signals are displayed in the front-end.

Here is a visualization of how one can represent a signal, and which direction it protects.

![Signal direction example](../svg_diagrams/signal_dir.en.svg)

<br />

The way the signals are arranged in an infrastructure depends on the local convention and can vary a lot from region to region.

Here is the convention we adopted in our example infrastructure:

> * We add two signals for each detector that are simply cutting the track sections into smaller TVDS, one in each direction.

Reviewer: ```suggestion
* We add two spacing signals (one per direction) for each detector that is cutting a long TVD section into smaller ones.
```

> * For detectors marking the entry of switches, we add only one signal, visible in the direction from which you enter the switch. Indeed, we chose not to be able to stop trains inside switches, because a train stopped in a switch makes it unusable by other trains.

Reviewer: ```suggestion
* Switch entries where a train might have to stop are protected by a signal (which is located outside of the switch TVD section). It must be visible from the direction used to approach the switch. When there are multiple switches in a row, only the first one usually needs protection, as interlocking is usually designed as not to encourage trains stopping in the middle of intersections.
```

* For the detectors that are in between two switches, we do not add signals, for the same reason as before.

> And there we have all the signals placed. Please note that we do not represent detectors that are linked to at least one signal.

Reviewer: ```suggestion
Note that detectors linked to at least one signal are not represented, as there are not signals without associated detectors in this example.
```

Is my proposed change correct? It's not always the case in real life

Author: It is

To get the `id` of a detector linked to a signal, take the signal's `id` and replace *S* by *D* (e.g. SA0 -> DA0).

![Infra diagram with all signals](../svg_diagrams/small_infra_signals.drawio.en.svg)

{{% alert title="" color="info"%}}

> On *TA6*, *TA7*, *TD0* and *TD1* we could not represent all signals because these track sections have a large `length` attribute and bear many detectors, hence many signals.

Reviewer: ```suggestion
On *TA6*, *TA7*, *TD0* and *TD1* we could not represent all signals because these track sections are very long and have many detectors, hence many signals.
```
